src/sentry/spans/consumers/process/factory.py

```diff
@@ -0,0 +1,55 @@
+import logging
+from collections.abc import Mapping
+from typing import Any
+
+from arroyo.backends.kafka.consumer import KafkaPayload
+from arroyo.processing.strategies.abstract import ProcessingStrategy, ProcessingStrategyFactory
+from arroyo.processing.strategies.commit import CommitOffsets
+from arroyo.processing.strategies.run_task import RunTask
+from arroyo.types import BrokerValue, Commit, Message, Partition
+from sentry_kafka_schemas import get_codec
+from sentry_kafka_schemas.codecs import Codec
+from sentry_kafka_schemas.schema_types.snuba_spans_v1 import SpanEvent
+
+from sentry.spans.buffer.redis import RedisSpansBuffer
+from sentry.tasks.spans import process_segment
+
+logger = logging.getLogger(__name__)
+SPAN_SCHEMA: Codec[SpanEvent] = get_codec("snuba-spans")
+
+PROCESS_SEGMENT_DELAY = 2 * 60  # 2 minutes
+
+
+def _deserialize_span(value: bytes) -> Mapping[str, Any]:
+    return SPAN_SCHEMA.decode(value)
+
+
+def process_message(message: Message[KafkaPayload]):
+    assert isinstance(message.value, BrokerValue)
+    try:
+        span = _deserialize_span(message.payload.value)
+        segment_id = span["segment_id"]
+        project_id = span["project_id"]
+    except Exception:
+        logger.exception("Failed to process span payload")
+
+    client = RedisSpansBuffer()
+    new_segment = client.write_span(project_id, segment_id, message.payload.value)
+    if new_segment:
+        # This function currently does nothing.
+        process_segment.apply_async(
+            args=[project_id, segment_id],
+            countdown=PROCESS_SEGMENT_DELAY,
+        )
```

Comment:
Did you want some kind of deduping here? Otherwise, you could create up to 1000 (maybe this limit changed, idk) tasks per segment.

Reply:
we are only creating a task the first time we see a segment id